Paper: Implementation of Code Editor (#323)
PaulBredl/config-assistant: paper/implementation.tex

\subsection{Code Editor}\label{subsec:code-editor}

> The code editor is a text editor program designed for editing the \cfgfiles.

Reviewer: ```suggestion
The code editor is a GUI panel designed for editing the \cfgfiles.
```

It can be easily embedded in any web page or JavaScript application.

\subsubsection{Features}

To make our code editor more user-friendly, we implemented several features, which are described in the following.

\paragraph{Schema Validation}
To ensure that the data is valid as soon as it is entered by the user, we use schema validation.
Therefore, the config data get validated based on the schema that is currently loaded.
We choose to use \textbf{Ajv JSON schema validator} \cite{ajv-validator}.
It supports the newest JSON schema draft 2020-12.
Ajv firstly generates a fast validation function from the input schema.
If the config data is not valid according to the schema, there will be a red error hint on left side of the code editor.

\paragraph{Support for JSON and YAML}
There is a lot build-in support for different languages in the ace editor.
In our project, we support JSON and YAML. It can be switched in the Settings page.

\paragraph{Concrete Syntax Tree}
% TODO: how we get cursor position and path correctly using cst

\paragraph{Syntax Highlighting}
The ace editor supports syntax highlighting for both JSON and YAML, which we use.

> With this functionality, users could better recognize different components in the data, like properties or values.

Reviewer: ```suggestion
With this functionality, users can recognize different components in the data better, like properties or values.
```

\subsection{Code Editor}\label{subsec:code-editor}

> The code editor is a text editor program designed for editing source code of computer programs.

Reviewer: This doesn't make a lot of sense. Our code editor does not allow editing source code.
```suggestion
The code editor is a text editor program designed for editing the \cfgfiles.
```

> In our user case, we use the very popular and strong editor called  \textbf{Ace Editor}. \cite{Ace-Editor}

Reviewer: ```suggestion
In our user case, we use the popular editor \textbf{Ace Editor}. \cite{Ace-Editor}
```

It is a browser based editor that matches and extends the features, usability and performance of existing native editors such as TextMate, Vim or Eclipse.
It can be easily embedded in any web page or JavaScript application.

\subsubsection{Features}

> In order to make our code editor more user-friendly, we implemented several features, which are described in the following.

Reviewer: ```suggestion
To make our code editor more user-friendly, we implemented several features, which are described in the following.
```

\paragraph{Schema Validation}

> In order to ensure that our data is valid as soon as it's received.

Reviewer: This isn't a full sentence..
```suggestion
To ensure that the data is valid as soon as it is entered by the user, we use schema validation.
```

> We need to have our config data get validated based on our schema.

Reviewer: Therefore, 
```suggestion
Therefore, the config data get validated based on the schema that is currently loaded.
```

> We choose to use a useful tool named \textbf{Ajv JSON schema validator} \cite{ajv-validator}.

Reviewer: ```suggestion
We choose to use \textbf{Ajv JSON schema validator} \cite{ajv-validator}.
```

> It could naturally support the newest JSON schema draft 2020-12.

Reviewer: ```suggestion
It supports the newest JSON schema draft 2020-12.
```

Ajv firstly generates the input JSON Schemas into super-fast validation functions.

> If config data is violated against the schema, there will be a small error hint on left side.

Reviewer: ```suggestion
If the config data is not valid according to the the schema, there will be a red error hint on left side of the code editor.
```
Todo consider screenshit

> Ajv also takes advantage of TypeScript type system to provide additional functionality.

Reviewer: ```suggestion
```
which we don't use right? So no need to mention it

> In our case, we have a helper class for the validation service, so that it's more convenient to deal with validation.

Reviewer: ```suggestion
```
I would leave that out

\paragraph{Support for JSON and YAML}

> The ace editor is very compatible and there're a lot build-in support for different languages.

Reviewer: ```suggestion
There is a lot build-in support for different languages in the ace editor.
```

In our project, we mainly support JSON and YAML. It could be switched in the Settings page.

\paragraph{Concrete Syntax Tree}
% TODO: how we get cursor position and path correctly using cst

\paragraph{Syntax Highlighting}
We have also implemented syntax highlighting for both JSON and YAML.
With this functionality, users could better recognize different components in the data, like properties or values.

\paragraph{Editor Operations}

> There are many other useful functionalities we have implemented in the code editor.

Reviewer: This is not done
